utils/firestore.js

```diff
@@ -1309,35 +1309,83 @@ const updateTempCheckDate = async (institute) => {
-    const snapshot = await db
-      .collection('boxes')
-      .where('132929440', '==', boxId)
-      .where('789843387', '==', siteCode)
-      .get();
+    try {
+        const getSnapshot = (query) => {
+            if (transaction) {
+                return transaction.get(query);
+            } else {
+                return query.get();
+            }
+        };
 
-    if (snapshot.size !== 1) {
-      return false;
+        if (boxIdArray.length <= 30 && isBPTL) {
+            const snapshot = await getSnapshot(db.collection('boxes').where(shippingBoxId, 'in', boxIdArray));
+            return snapshot.docs.map(document => ({ data: document.data(), docRef: document.ref }));
+        } else if (boxIdArray.length <= 15 && !isBPTL) {
+            const snapshot = await getSnapshot(db.collection('boxes').where(shippingBoxId, 'in', boxIdArray).where(loginSite, '==', siteCode));
+            return snapshot.docs.map(document => ({ data: document.data(), docRef: document.ref }));
+        } else {
+            const queries = boxIdArray.map(boxId => {
+                return db.collection('boxes').where(shippingBoxId, '==', boxId).where(loginSite, '==', siteCode);
+            });
+        
+            const boxResults = await Promise.all(queries.map(query => getSnapshot(query)));
+            return boxResults.map(result => result.docs.map(document => ({ data: document.data(), docRef: document.ref }))).flat();
+        }
+    } catch (error) {
+        throw new Error(error);
     }
+}
 
-    batch.update(snapshot.docs[0].ref, shipmentData);
-  }
+/**
+ * Ship a batch of boxes using a transaction
+ * Transaction: (1) Guarantees atomicity (2) Ensures data integrity (3) Automatically retries on initial failure
+ * @param {Array} boxIdAndShipmentDataArray - array of objects with boxId and shipmentData
+ * @param {number} siteCode - site code of the user (number)
+ * @returns {boolean} true if successful, throws error otherwise
+ */
+const shipBatchBoxes = async (boxIdAndShipmentDataArray, siteCode) => {
+    const boxIdToShipmentData = {};
+    boxIdAndShipmentDataArray.forEach(item => {
+        boxIdToShipmentData[item.boxId] = item.shipmentData;
+    });
+
+    const boxIdArray = Object.keys(boxIdToShipmentData);
 
-  await batch.commit().catch((err) => {
-    console.log('Error occurred when commiting box data:\n', err);
-    return false;
-  });
+    try {
+        await db.runTransaction(async (transaction) => {
+            const boxes = await getBoxesByBoxId(boxIdArray, siteCode, false, transaction);
+    
+            for (const box of boxes) {
+                const boxData = box.data;
+                const shipmentData = boxIdToShipmentData[boxData[fieldMapping.shippingBoxId]];
+
+                if (shipmentData) {
+                    shipmentData[fieldMapping.submitShipmentFlag] = fieldMapping.yes;
+                    Object.assign(boxData, shipmentData);
+                    transaction.update(box.docRef, boxData);
```

Comment:
Initially, I was curious why you were using map to transform each document to `{ data: document.data(), docRef: document.ref }` in the `getBoxesByBoxId` function control flow. This line of code explains why you needed that document.ref, it's for updating. Also, I never used or seen `db.runTransaction`. Thanks for the jsdocs and using it so I can read up on it.

Reply:
Thanks, Gene. Normally, I think the previous batch write was substantial enough for this operation. Since this has been a persistent issue, I like the idea of using a transaction here and tightening up the entire call chain end-to-end. It's a little slower, but we're likely talking about milliseconds. I like that transactions auto-retry and lock access to the associated docs while the transaction runs...it's just an extra layer of certainty that our shipped box data will be consistent.

Reply:
The problem is that, based on my checks on thousands lines of logs, there were no backend errors in those failed shipments: `ship` API was not called. I believe the problem lies in frontend code.